[Bug] [hudi sink] hudi sink java client not support struct field type where is null

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

INSERT INTO test.dcjob_hudi2hudi_ly_0122_1 (`_hoodie_commit_time`, `_hoodie_commit_seqno`, `_hoodie_record_key`, `_hoodie_partition_path`, `_hoodie_file_name`, int_column, string_column, varchar_column, tinyint_column, smallint_column, bigint_column, float_column, double_column, decimal_column, date_column, timestamp_column, char_column, boolean_column, binary_column, array_column, map_column, struct_column) VALUES('20260306093746536', '20260306093746536_0_6', '11', '', 'eb787520-2452-4f83-b623-4b89c89b29e2-0_0-4733-5267_20260306154907638.parquet', 11, 'varchar_column1', 'string值1', 121, 321, 922337203685477501, 1.21, 123.41, 12345678.91, '2026-01-01', '2026-01-01 15:30:20.000', 'char值1', false, 0x62696E6172795F636F6C756D6E31, NULL, NULL, []);

Caused by: java.lang.NullPointerException
	at org.apache.avro.generic.GenericData.getField(GenericData.java:846)
	at org.apache.avro.generic.GenericData.getField(GenericData.java:865)
	at org.apache.avro.generic.GenericDatumWriter.writeField(GenericDatumWriter.java:219)
	at org.apache.avro.generic.GenericDatumWriter.writeRecord(GenericDatumWriter.java:210)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:131)
	at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:83)
	at org.apache.avro.generic.GenericDatumWriter.writeField(GenericDatumWriter.java:221)
	at org.apache.avro.generic.GenericDatumWriter.writeRecord(GenericDatumWriter.java:210)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:131)

### SeaTunnel Version

2.3.9

### SeaTunnel Config

```conf
{
	"env": {
		"job.mode": "BATCH",
		"parallelism": 1,
		"job.name": "zsdcTest_hudi2hudi_0121_instance_1768987020918_5401",
		"checkpoint.interval": 10
	},
	"source": [{
		"_type": "hudi_source",
		"plugin_name": "hudi",
		"result_table_name": "dcjob_hudi2hudi_ly_0121_12101424056",
		"table_dfs_path": "hdfs://nameservice1/user/hive/warehouse",
		"table_name": "dcjob_hudi2hudi_ly_0122_1",
		"database": "test",
		"query": "select * from `dcjob_hudi2hudi_ly_0122_1`"
	}],
	"sink": [{
		"_type": "hudi_sink",
		"schema_save_mode": "ERROR_WHEN_SCHEMA_NOT_EXIST",
		"data_save_mode": "APPEND_DATA",
		"plugin_name": "Hudi",
		"source_table_name": "dcjob_hudi2hudi_ly_0121_12101424056_t",
		"table_dfs_path": "hdfs://nameservice1/user/hive/warehouse",
		"table_name": "dcjob_hudi2hudi_mb_0122_1",
		"database": "test",
		"table_type": "COPY_ON_WRITE"
	}],
	"transform": [{
		"_type": "sql_trans",
		"query": "SELECT TRY_CAST(`int_column` AS INTEGER) as `int_column`,TRY_CAST(`string_column` AS STRING) as `string_column`,TRY_CAST(`varchar_column` AS STRING) as `varchar_column`,TRY_CAST(`tinyint_column` AS INTEGER) as `tinyint_column`,TRY_CAST(`smallint_column` AS INTEGER) as `smallint_column`,TRY_CAST(`bigint_column` AS BIGINT) as `bigint_column`,TRY_CAST(`float_column` AS FLOAT) as `float_column`,TRY_CAST(`double_column` AS DOUBLE) as `double_column`,TRY_CAST(`decimal_column` AS DECIMAL(10,2)) as `decimal_column`,TRY_CAST(`date_column` AS DATE) as `date_column`,TRY_CAST(`timestamp_column` AS TIMESTAMP) as `timestamp_column`,TRY_CAST(`char_column` AS STRING) as `char_column`,TRY_CAST(`boolean_column` AS BOOLEAN) as `boolean_column`,TRY_CAST(`binary_column` AS BYTES) as `binary_column`,TRY_CAST(`array_column` AS ARRAY<STRING>) as `array_column`,TRY_CAST(`map_column` AS MAP(VARCHAR, INT)) as `map_column`,TRY_CAST(`struct_column` AS ROW(field1 string,field2 int)) as `struct_column` FROM `dcjob_hudi2hudi_ly_0121_12101424056`",
		"source_table_name": "dcjob_hudi2hudi_ly_0121_12101424056",
		"result_table_name": "dcjob_hudi2hudi_ly_0121_12101424056_t"
	}]
}
```

### Running Command

```shell
xxx
```

### Error Exception

```log
Caused by: java.lang.NullPointerException
	at org.apache.avro.generic.GenericData.getField(GenericData.java:846)
	at org.apache.avro.generic.GenericData.getField(GenericData.java:865)
	at org.apache.avro.generic.GenericDatumWriter.writeField(GenericDatumWriter.java:219)
	at org.apache.avro.generic.GenericDatumWriter.writeRecord(GenericDatumWriter.java:210)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:131)
	at org.apache.avro.generic.GenericDatumWriter.write(GenericDatumWriter.java:83)
	at org.apache.avro.generic.GenericDatumWriter.writeField(GenericDatumWriter.java:221)
	at org.apache.avro.generic.GenericDatumWriter.writeRecord(GenericDatumWriter.java:210)
	at org.apache.avro.generic.GenericDatumWriter.writeWithoutConversion(GenericDatumWriter.java:131)
```

### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

case ROW:
                SeaTunnelRowType rowType = (SeaTunnelRowType) dataType;
                List<String> fieldNames = Arrays.asList(rowType.getFieldNames());
                // we have to make sure the record name is different in a Schema
                SchemaBuilder.FieldAssembler<Schema> builder =
                        SchemaBuilder.builder().record(rowName).fields();
                for (int i = 0; i < fieldNames.size(); i++) {
                    String fieldName = fieldNames.get(i);

                    SeaTunnelDataType<?> fieldType = rowType.getFieldType(i);
                    SchemaBuilder.GenericDefault<Schema> fieldBuilder =
                            builder.name(fieldName)
                                    .type(convertToSchema(fieldType, rowName + "." + fieldName));

                    //struct row类型，默认值不能之间设置为null
                    builder = fieldBuilder.noDefault();

                }
                return nullableSchema(builder.endRecord());

**加了nullableSchema还是报错**
Exception in thread "main" org.apache.spark.SparkException: Writing job aborted
	at org.apache.spark.sql.errors.QueryExecutionErrors$.writingJobAbortedError(QueryExecutionErrors.scala:749)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:409)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2$(WriteToDataSourceV2Exec.scala:353)
	at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.writeWithV2(WriteToDataSourceV2Exec.scala:244)
	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run(WriteToDataSourceV2Exec.scala:332)
	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run$(WriteToDataSourceV2Exec.scala:331)
	at org.apache.spark.sql.execution.datasources.v2.AppendDataExec.run(WriteToDataSourceV2Exec.scala:244)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43)
	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:109)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:95)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:779)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:98)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:94)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:176)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:584)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:560)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:94)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:81)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:79)
	at org.apache.spark.sql.execution.QueryExecution.assertCommandExecuted(QueryExecution.scala:116)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:860)
	at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:311)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:247)
	at org.apache.seatunnel.core.starter.spark.execution.SinkExecuteProcessor.execute(SinkExecuteProcessor.java:131)
	at org.apache.seatunnel.core.starter.spark.execution.SparkExecution.execute(SparkExecution.java:72)
	at com.unionbigdata.rdc.seatunnel.spark.SeatunnelSparkStarter.main(SeatunnelSparkStarter.java:35)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (DESKTOP-UTQRQFB executor driver): org.apache.seatunnel.connectors.seatunnel.hudi.exception.HudiConnectorException: ErrorCode:[COMMON-11], ErrorDescription:[Sink writer operation failed, such as (open, close) etc...] - Writing records to Hudi failed.
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.writeRecord(HudiRecordWriter.java:127)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiSinkWriter.write(HudiSinkWriter.java:84)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiSinkWriter.write(HudiSinkWriter.java:39)
	at org.apache.seatunnel.translation.spark.sink.write.SeaTunnelSparkDataWriter.write(SeaTunnelSparkDataWriter.java:69)
	at org.apache.seatunnel.translation.spark.sink.write.SeaTunnelSparkDataWriter.write(SeaTunnelSparkDataWriter.java:40)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$1(WriteToDataSourceV2Exec.scala:442)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1538)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:480)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:381)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: org.apache.avro.AvroRuntimeException: Not a record schema: ["null",{"type":"record","name":"dcjob_hudi2hudi_mb_0122_1_record","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1","fields":[{"name":"int_column","type":["null","int"]},{"name":"string_column","type":["null","string"]},{"name":"varchar_column","type":["null","string"]},{"name":"tinyint_column","type":["null","int"]},{"name":"smallint_column","type":["null","int"]},{"name":"bigint_column","type":["null","long"]},{"name":"float_column","type":["null","float"]},{"name":"double_column","type":["null","double"]},{"name":"decimal_column","type":["null",{"type":"fixed","name":"fixed","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1.dcjob_hudi2hudi_mb_0122_1_record.decimal_column","size":5,"logicalType":"decimal","precision":10,"scale":2}]},{"name":"date_column","type":["null",{"type":"int","logicalType":"date"}]},{"name":"timestamp_column","type":["null",{"type":"long","logicalType":"timestamp-millis"}]},{"name":"char_column","type":["null","string"]},{"name":"boolean_column","type":["null","boolean"]},{"name":"binary_column","type":["null","bytes"]},{"name":"array_column","type":["null",{"type":"array","items":["null","string"]}]},{"name":"map_column","type":["null",{"type":"map","values":["null","int"]}]},{"name":"struct_column","type":["null",{"type":"record","name":"struct_column","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1.dcjob_hudi2hudi_mb_0122_1_record","fields":[{"name":"field1","type":["null","string"]},{"name":"field2","type":["null","int"]}]}]}]}]
	at org.apache.avro.generic.GenericData$Record.<init>(GenericData.java:227)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.convert.HudiRecordConverter.convertRow(HudiRecordConverter.java:66)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.prepareRecords(HudiRecordWriter.java:200)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.writeRecord(HudiRecordWriter.java:120)
	... 16 more

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2672)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2608)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2607)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2607)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1182)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1182)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1182)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2860)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2802)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2791)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:952)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2228)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:377)
	... 35 more
Caused by: org.apache.seatunnel.connectors.seatunnel.hudi.exception.HudiConnectorException: ErrorCode:[COMMON-11], ErrorDescription:[Sink writer operation failed, such as (open, close) etc...] - Writing records to Hudi failed.
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.writeRecord(HudiRecordWriter.java:127)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiSinkWriter.write(HudiSinkWriter.java:84)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiSinkWriter.write(HudiSinkWriter.java:39)
	at org.apache.seatunnel.translation.spark.sink.write.SeaTunnelSparkDataWriter.write(SeaTunnelSparkDataWriter.java:69)
	at org.apache.seatunnel.translation.spark.sink.write.SeaTunnelSparkDataWriter.write(SeaTunnelSparkDataWriter.java:40)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$1(WriteToDataSourceV2Exec.scala:442)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1538)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:480)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:381)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:136)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:548)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1504)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:551)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: org.apache.avro.AvroRuntimeException: Not a record schema: ["null",{"type":"record","name":"dcjob_hudi2hudi_mb_0122_1_record","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1","fields":[{"name":"int_column","type":["null","int"]},{"name":"string_column","type":["null","string"]},{"name":"varchar_column","type":["null","string"]},{"name":"tinyint_column","type":["null","int"]},{"name":"smallint_column","type":["null","int"]},{"name":"bigint_column","type":["null","long"]},{"name":"float_column","type":["null","float"]},{"name":"double_column","type":["null","double"]},{"name":"decimal_column","type":["null",{"type":"fixed","name":"fixed","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1.dcjob_hudi2hudi_mb_0122_1_record.decimal_column","size":5,"logicalType":"decimal","precision":10,"scale":2}]},{"name":"date_column","type":["null",{"type":"int","logicalType":"date"}]},{"name":"timestamp_column","type":["null",{"type":"long","logicalType":"timestamp-millis"}]},{"name":"char_column","type":["null","string"]},{"name":"boolean_column","type":["null","boolean"]},{"name":"binary_column","type":["null","bytes"]},{"name":"array_column","type":["null",{"type":"array","items":["null","string"]}]},{"name":"map_column","type":["null",{"type":"map","values":["null","int"]}]},{"name":"struct_column","type":["null",{"type":"record","name":"struct_column","namespace":"hoodie.dcjob_hudi2hudi_mb_0122_1.dcjob_hudi2hudi_mb_0122_1_record","fields":[{"name":"field1","type":["null","string"]},{"name":"field2","type":["null","int"]}]}]}]}]
	at org.apache.avro.generic.GenericData$Record.<init>(GenericData.java:227)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.convert.HudiRecordConverter.convertRow(HudiRecordConverter.java:66)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.prepareRecords(HudiRecordWriter.java:200)
	at org.apache.seatunnel.connectors.seatunnel.hudi.sink.writer.HudiRecordWriter.writeRecord(HudiRecordWriter.java:120)
	... 16 more

--- a/seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/convert/AvroSchemaConverter.java
+++ b/seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/convert/AvroSchemaConverter.java
@@ -64,6 +64,11 @@ public class AvroSchemaConverter implements Serializable {
      * @return Avro's {@link Schema} matching this logical type.
      */
     public static Schema convertToSchema(SeaTunnelDataType<?> dataType, String rowName) {
+        return convertToSchema(dataType, rowName, false);
+    }
+
+    private static Schema convertToSchema(
+            SeaTunnelDataType<?> dataType, String rowName, boolean nullableRow) {
         switch (dataType.getSqlType()) {
             case BOOLEAN:
                 Schema bool = SchemaBuilder.builder().booleanType();
@@ -126,25 +131,30 @@ public class AvroSchemaConverter implements Serializable {
                     SeaTunnelDataType<?> fieldType = rowType.getFieldType(i);
                     SchemaBuilder.GenericDefault<Schema> fieldBuilder =
                             builder.name(fieldName)
-                                    .type(convertToSchema(fieldType, rowName + "." + fieldName));
+                                    .type(
+                                            convertToSchema(
+                                                    fieldType, rowName + "." + fieldName, true));
 
                     builder = fieldBuilder.withDefault(null);
                 }
-                return builder.endRecord();
+                Schema record = builder.endRecord();
+                return nullableRow ? nullableSchema(record) : record;
             case MAP:
                 Schema map =
                         SchemaBuilder.builder()
                                 .map()
                                 .values(
                                         convertToSchema(
-                                                extractValueTypeToAvroMap(dataType), rowName));
+                                                extractValueTypeToAvroMap(dataType),
+                                                rowName,
+                                                true));
                 return nullableSchema(map);
             case ARRAY:
                 ArrayType<?, ?> arrayType = (ArrayType<?, ?>) dataType;
                 Schema array =
                         SchemaBuilder.builder()
                                 .array()
-                                .items(convertToSchema(arrayType.getElementType(), rowName));
+                                .items(convertToSchema(arrayType.getElementType(), rowName, true));
                 return nullableSchema(array);
             default:
                 throw new UnsupportedOperationException(

--- a/seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/convert/HudiRecordConverter.java
+++ b/seatunnel-connectors-v2/connector-hudi/src/main/java/org/apache/seatunnel/connectors/seatunnel/hudi/sink/convert/HudiRecordConverter.java
@@ -24,7 +24,6 @@ import org.apache.seatunnel.connectors.seatunnel.hudi.config.HudiTableConfig;
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.generic.GenericRecord;
-import org.apache.hudi.avro.AvroSchemaUtils;
 import org.apache.hudi.common.model.HoodieAvroPayload;
 import org.apache.hudi.common.model.HoodieAvroRecord;
 import org.apache.hudi.common.model.HoodieKey;
@@ -39,7 +38,6 @@ import java.util.Arrays;
 import java.util.UUID;
 import java.util.stream.Collectors;
 
-import static org.apache.seatunnel.connectors.seatunnel.hudi.sink.convert.AvroSchemaConverter.convertToSchema;
 import static org.apache.seatunnel.connectors.seatunnel.hudi.sink.convert.RowDataToAvroConverters.createConverter;
 
 public class HudiRecordConverter implements Serializable {
@@ -59,17 +57,11 @@ public class HudiRecordConverter implements Serializable {
             HudiTableConfig hudiTableConfig) {
         GenericRecord rec = new GenericData.Record(schema);
         for (int i = 0; i < seaTunnelRowType.getTotalFields(); i++) {
+            String fieldName = seaTunnelRowType.getFieldNames()[i];
             rec.put(
-                    seaTunnelRowType.getFieldNames()[i],
+                    fieldName,
                     createConverter(seaTunnelRowType.getFieldType(i))
-                            .convert(
-                                    convertToSchema(
-                                            seaTunnelRowType.getFieldType(i),
-                                            AvroSchemaUtils.getAvroRecordQualifiedName(
-                                                            hudiTableConfig.getTableName())
-                                                    + "."
-                                                    + seaTunnelRowType.getFieldNames()[i]),
-                                    element.getField(i)));
+                            .convert(schema.getField(fieldName).schema(), element.getField(i)));
         }
         return new HoodieAvroRecord<>(
                 getHoodieKey(element, seaTunnelRowType, hudiTableConfig),

--- a/seatunnel-connectors-v2/connector-hudi/src/test/java/org/apache/seatunnel/connectors/seatunnel/hudi/HudiTest.java
+++ b/seatunnel-connectors-v2/connector-hudi/src/test/java/org/apache/seatunnel/connectors/seatunnel/hudi/HudiTest.java
@@ -23,10 +23,15 @@ import org.apache.seatunnel.api.table.type.MapType;
 import org.apache.seatunnel.api.table.type.SeaTunnelDataType;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
 import org.apache.seatunnel.api.table.type.SeaTunnelRowType;
+import org.apache.seatunnel.connectors.seatunnel.hudi.config.HudiTableConfig;
+import org.apache.seatunnel.connectors.seatunnel.hudi.sink.convert.HudiRecordConverter;
 
 import org.apache.avro.Schema;
 import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericDatumWriter;
 import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.EncoderFactory;
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hudi.avro.AvroSchemaUtils;
 import org.apache.hudi.client.HoodieJavaWriteClient;
@@ -37,6 +42,7 @@ import org.apache.hudi.common.model.HoodieAvroRecord;
 import org.apache.hudi.common.model.HoodieKey;
 import org.apache.hudi.common.model.HoodieRecord;
 import org.apache.hudi.common.model.HoodieTableType;
+import org.apache.hudi.common.model.WriteOperationType;
 import org.apache.hudi.common.table.HoodieTableMetaClient;
 import org.apache.hudi.common.util.Option;
 import org.apache.hudi.common.util.StringUtils;
@@ -53,6 +59,7 @@ import org.junit.jupiter.api.condition.DisabledOnOs;
 import org.junit.jupiter.api.condition.OS;
 import org.junit.jupiter.api.io.TempDir;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.math.BigDecimal;
 import java.sql.Timestamp;
@@ -125,6 +132,45 @@ public class HudiTest {
         Assertions.assertEquals(
                 "{\"type\":\"record\",\"name\":\"hudi_record\",\"namespace\":\"hoodie.hudi\",\"fields\":[{\"name\":\"bool\",\"type\":[\"null\",\"boolean\"],\"default\":null},{\"name\":\"int\",\"type\":[\"null\",\"int\"],\"default\":null},{\"name\":\"longValue\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"float\",\"type\":[\"null\",\"float\"],\"default\":null},{\"name\":\"name\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"date\",\"type\":[\"null\",{\"type\":\"int\",\"logicalType\":\"date\"}],\"default\":null},{\"name\":\"time\",\"type\":[\"null\",{\"type\":\"int\",\"logicalType\":\"time-millis\"}],\"default\":null},{\"name\":\"timestamp3\",\"type\":[\"null\",{\"type\":\"long\",\"logicalType\":\"timestamp-millis\"}],\"default\":null},{\"name\":\"map\",\"type\":[\"null\",{\"type\":\"map\",\"values\":[\"null\",\"long\"]}],\"default\":null},{\"name\":\"decimal\",\"type\":[\"null\",{\"type\":\"fixed\",\"name\":\"fixed\",\"namespace\":\"hoodie.hudi.hudi_record.decimal\",\"size\":5,\"logicalType\":\"decimal\",\"precision\":10,\"scale\":5}],\"default\":null}]}",
                 getSchema());
+    }
+
+    @Test
+    void testConvertNullableStructField() throws IOException {
+        SeaTunnelRowType structType =
+                new SeaTunnelRowType(
+                        new String[] {"field1", "field2"},
+                        new SeaTunnelDataType[] {STRING_TYPE, INT_TYPE});
+        SeaTunnelRowType rowType =
+                new SeaTunnelRowType(
+                        new String[] {"id", "payload"},
+                        new SeaTunnelDataType[] {INT_TYPE, structType});
+        Schema schema =
+                convertToSchema(rowType, AvroSchemaUtils.getAvroRecordQualifiedName(tableName));
+
+        Assertions.assertEquals(Schema.Type.RECORD, schema.getType());
+        Assertions.assertEquals(Schema.Type.UNION, schema.getField("payload").schema().getType());
+
+        SeaTunnelRow nullStructRow = new SeaTunnelRow(2);
+        nullStructRow.setField(0, 1);
+        nullStructRow.setField(1, null);
+
+        GenericRecord nullStructRecord = convertToAvroRecord(schema, rowType, nullStructRow);
+        Assertions.assertNull(nullStructRecord.get("payload"));
+        writeAvroRecord(schema, nullStructRecord);
+
+        SeaTunnelRow struct = new SeaTunnelRow(2);
+        struct.setField(0, "value");
+        struct.setField(1, 100);
+
+        SeaTunnelRow structRow = new SeaTunnelRow(2);
+        structRow.setField(0, 2);
+        structRow.setField(1, struct);
+
+        GenericRecord structRecord = convertToAvroRecord(schema, rowType, structRow);
+        GenericRecord payload = (GenericRecord) structRecord.get("payload");
+        Assertions.assertEquals("value", payload.get("field1").toString());
+        Assertions.assertEquals(100, payload.get("field2"));
+        writeAvroRecord(schema, structRecord);
     }
 
     @Test
@@ -208,6 +254,25 @@ public class HudiTest {
 
         return new HoodieAvroRecord<>(
                 getHoodieKey(element, seaTunnelRowType), new HoodieAvroPayload(Option.of(rec)));
+    }
+
+    private GenericRecord convertToAvroRecord(
+            Schema schema, SeaTunnelRowType rowType, SeaTunnelRow element) throws IOException {
+        HudiTableConfig tableConfig =
+                HudiTableConfig.builder()
+                        .tableName(tableName)
+                        .opType(WriteOperationType.INSERT)
+                        .build();
+        HoodieRecord<HoodieAvroPayload> record =
+                new HudiRecordConverter().convertRow(schema, rowType, element, tableConfig);
+        return (GenericRecord) record.getData().getInsertValue(schema).get();
+    }
+
+    private void writeAvroRecord(Schema schema, GenericRecord record) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(out, null);
+        new GenericDatumWriter<GenericRecord>(schema).write(record, encoder);
+        encoder.flush();
     }
 
     private HoodieKey getHoodieKey(SeaTunnelRow element, SeaTunnelRowType seaTunnelRowType) {

